File: code/richediting/src/main/java/org/fbme/ide/richediting/adapters/fbnetwork/actions/ExpandAction.java
package org.fbme.ide.richediting.adapters.fbnetwork.actions;

import com.intellij.openapi.util.Pair;
import jetbrains.mps.editor.runtime.HeadlessEditorComponent;
import jetbrains.mps.openapi.editor.cells.EditorCell;
import jetbrains.mps.openapi.editor.style.Style;
import org.fbme.ide.iec61499.repository.PlatformElement;
import org.fbme.ide.richediting.adapters.fbnetwork.ExpandedComponentsController;
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionPath;
import org.fbme.ide.richediting.adapters.fbnetwork.FunctionBlockController;
import org.fbme.ide.richediting.viewmodel.FunctionBlockView;
import org.fbme.ide.richediting.viewmodel.NetworkComponentView;
import org.fbme.ide.richediting.viewmodel.NetworkConnectionView;
import org.fbme.ide.richediting.viewmodel.NetworkPortView;
import org.fbme.lib.iec61499.fbnetwork.ConnectionPath;
import org.fbme.lib.iec61499.instances.FunctionBlockInstance;
import org.fbme.lib.iec61499.instances.Instance;
import org.fbme.scenes.controllers.LayoutUtil;
import org.fbme.scenes.controllers.diagram.PortController;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.model.SNode;
import org.jetbrains.mps.openapi.module.SRepository;

import java.awt.*;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ExpandAction extends ExpandOrCollapseAction {

    private final String EXPANDED_FB_INSTANCE_KEY = "org.fbme.ide.richediting.lang.editor.Rich Editing Hint.expanded_fb_instance";

    public ExpandAction(EditorCell cell) {
        super(cell);
    }

    @Override
    public void apply() {
        expand(selectedFBs);
    }

    private void expand(Set<NetworkComponentView> selectedComponents) {
        for (NetworkComponentView selectedComponent : selectedComponents) {
            if (selectedComponent instanceof FunctionBlockView) {
                expand((FunctionBlockView) selectedComponent);
            }
        }
    }

    private void expand(FunctionBlockView component) {
        FunctionBlockController componentController = (FunctionBlockController) componentsFacility.getController(component);

        Point editorComponentPosition = componentsFacility.getModelForm(component);
        Rectangle oldBounds = componentController.getFBCellBounds(editorComponentPosition);

        int lineSize = LayoutUtil.getLineSize(componentController.getComponentCell().getStyle());
        EditorCell sceneCell = createExpandedSceneCell(componentController);
        Rectangle newBounds = new Rectangle(oldBounds.x, oldBounds.y, sceneCell.getWidth(), sceneCell.getHeight() + lineSize);

        int dx = newBounds.width - oldBounds.width;
        int dy = newBounds.height - oldBounds.height;

        ExpandedComponentsController expandedComponentsController = componentController.getExpandedComponentsController();
        expandedComponentsController.addExpandedComponent(component, viewpoint.fromEditorDimension(dx), viewpoint.fromEditorDimension(dy));
        expandedComponentsController.addAffectedSections(component, getAffectedSections(oldBounds, component));
        expandedComponentsController.addAffectedComponents(component, getAffectedComponents(oldBounds));

        expandedComponentsController.update();
    }

    private Set<Pair<NetworkConnectionView, Integer>> getAffectedSections(Rectangle bounds, FunctionBlockView component) {
        Set<Pair<NetworkConnectionView, Integer>> affectedSections = new HashSet<>();

        int rightBound = bounds.x + bounds.width;
        int bottomBound = bounds.y + bounds.height;

        Set<NetworkConnectionView> connections = diagramController.getConnections();
        for (NetworkConnectionView connection : connections) {
            NetworkPortView source = diagramController.getSource(connection);
            NetworkPortView target = diagramController.getTarget(connection);

            PortController sourcePortController = diagramController.getPortController(source);
            PortController targetPortController = diagramController.getPortController(target);

            Point sourcePortPosition = sourcePortController.getModelEndpointPosition();
            Point targetPortPosition = targetPortController.getModelEndpointPosition();

            NetworkComponentView sourceComponent = diagramController.getComponent(source);
            NetworkComponentView targetComponent = diagramController.getComponent(target);

            Point sourceComponentPosition = componentsFacility.getModelForm(sourceComponent);
            Point targetComponentPosition = componentsFacility.getModelForm(targetComponent);

            FBConnectionPath path = connectionSynchronizer.getPath(connection).apply(sourcePortPosition, targetPortPosition);
            List<Point> bendPoints = path.getBendPoints();
            ConnectionPath.Kind pathKind = path.getPathKind();

            if (pathKind == ConnectionPath.Kind.MoreThanFour) {
                for (int i = 1; i < bendPoints.size(); i++) {
                    Point v = bendPoints.get(i);

                    boolean isHorizontal = i % 2 == 0;
                    if ((isHorizontal && v.y > bottomBound) || ((!isHorizontal) && v.x > rightBound)) {
                        affectedSections.add(new Pair<>(connection, i));
                    }
                }
            } else if (pathKind == ConnectionPath.Kind.TwoAngles) {
                int x1 = path.getX1();

                if (sourceComponent != component && sourceComponentPosition.x <= rightBound && bounds.x < x1) {
                    affectedSections.add(new Pair<>(connection, 1));
                }
            } else if (pathKind == ConnectionPath.Kind.FourAngles) {
                int x1 = path.getX1();
                int y = path.getY();
                int x2 = path.getX2();

                if (sourceComponent != component && sourceComponentPosition.x <= rightBound && bounds.x < x1) {
                    affectedSections.add(new Pair<>(connection, 1));
                }
                if (Math.min(sourcePortPosition.y, y) < bottomBound && bottomBound < Math.max(sourcePortPosition.y, y)) {
                    affectedSections.add(new Pair<>(connection, (sourcePortPosition.y < y ? 2 : -2)));
                }
                if (x2 < rightBound && rightBound < targetComponentPosition.x) {
                    affectedSections.add(new Pair<>(connection, -3));
                }
            }
        }
        return affectedSections;
    }

    @NotNull
    private EditorCell createExpandedSceneCell(FunctionBlockController componentController) {

        FunctionBlockInstance childInstance = componentController.getFbInstance();
        assert childInstance != null;
        Instance childNetworkInstance = childInstance.getContainedNetwork();
        assert childNetworkInstance != null;
        SNode node = ((PlatformElement) childInstance.getDeclaration()).getNode();
        SRepository repository = componentController.getComponentCell().getContext().getRepository();

        HeadlessEditorComponent editorComponent = new HeadlessEditorComponent(repository);
        editorComponent.getUpdater().setInitialEditorHints(new String[]{EXPANDED_FB_INSTANCE_KEY});
        editorComponent.editNode(node);
        EditorCell sceneCell = editorComponent.getRootCell();

        Style style = componentController.getComponentCell().getStyle();
        int fontSize = LayoutUtil.getFontSize(style);
        LayoutUtil.setFontSize(sceneCell.getStyle(), fontSize);
        sceneCell.relayout();

        return sceneCell;
    }

    private Pair<Set<NetworkComponentView>, Set<NetworkComponentView>> getAffectedComponents(Rectangle bounds) {
        Set<NetworkComponentView> affectedComponentsByX = new HashSet<>();
        Set<NetworkComponentView> affectedComponentsByY = new HashSet<>();

        int rightBound = bounds.x + bounds.width;
        int bottomBound = bounds.y + bounds.height;

        Set<NetworkComponentView> components = diagramController.getComponents();
        for (NetworkComponentView component : components) {
            Point modelComponentPosition = componentsFacility.getModelForm(component);

            if (modelComponentPosition.x > rightBound) {
                affectedComponentsByX.add(component);
            }
            if (modelComponentPosition.y > bottomBound) {
                affectedComponentsByY.add(component);
            }
        }
        return new Pair<>(affectedComponentsByX, affectedComponentsByY);
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/adapters/fbnetwork/actions/ExpandAction.kt
package org.fbme.ide.richediting.adapters.fbnetwork.actions

import jetbrains.mps.editor.runtime.HeadlessEditorComponent
import jetbrains.mps.openapi.editor.cells.EditorCell
import jetbrains.mps.openapi.editor.style.Style
import org.fbme.ide.iec61499.repository.PlatformElement
import org.fbme.ide.richediting.adapters.fbnetwork.FunctionBlockController
import org.fbme.ide.richediting.viewmodel.FunctionBlockView
import org.fbme.ide.richediting.viewmodel.InterfaceEndpointView
import org.fbme.ide.richediting.viewmodel.NetworkComponentView
import org.fbme.scenes.cells.EditorCell_Scene
import org.fbme.scenes.controllers.LayoutUtil.getFontSize
import org.fbme.scenes.controllers.LayoutUtil.getLineSize
import org.fbme.scenes.controllers.LayoutUtil.setFontSize
import org.jetbrains.mps.openapi.module.SRepository
import java.awt.Point
import java.awt.Rectangle

class ExpandAction(cell: EditorCell) : ExpandOrCollapseAction(cell) {
    fun apply() {
        val functionBlock = selectedFBs.filterIsInstance<FunctionBlockView>().last()
        functionBlock.expand()
    }

    private fun FunctionBlockView.expand() {
        val componentController = componentsFacility.getController(this) as FunctionBlockController
        val editorComponentPosition = componentsFacility.getModelForm(this)
        val oldBounds = componentController.getFBCellBounds(editorComponentPosition)
        val lineSize = getLineSize(componentController.componentCell.style)
        val sceneCell = createExpandedSceneCell(componentController)
        val bounds = (sceneCell as EditorCell_Scene).calculateBounds()

        val newBounds = Rectangle(
            oldBounds.x,
            oldBounds.y,
            EDITOR_SHIFT_X + bounds.width,
            EDITOR_SHIFT_Y + bounds.height + (1.5 * lineSize).toInt()
        )

        val dx = viewpoint.fromEditorDimension(newBounds.width - oldBounds.width)
        val dy = viewpoint.fromEditorDimension(newBounds.height - oldBounds.height)

        val expandedComponentsController = componentController.expandedComponentsController

        expandedComponentsController.addFB(
            functionBlock = this,
            editorShift = Point(EDITOR_SHIFT_X - bounds.x, EDITOR_SHIFT_Y - bounds.y),
            dx = dx,
            dy = dy,
            componentShifts(oldBounds, newBounds)
        )
    }

    private fun NetworkComponentView.componentShifts(
        oldBounds: Rectangle,
        newBounds: Rectangle
    ): MutableMap<NetworkComponentView, Point> {
        val componentShifts = mutableMapOf<NetworkComponentView, Point>()

        val rightBound = oldBounds.x + oldBounds.width
        val bottomBound = oldBounds.y + oldBounds.height
        val dx = viewpoint.fromEditorDimension(newBounds.width - oldBounds.width)
        val dy = viewpoint.fromEditorDimension(newBounds.height - oldBounds.height)

        val components = diagramController
            .components
            .filter { (it is FunctionBlockView || it is InterfaceEndpointView) && it != this }
        for (component in components) {
            val controller = componentsFacility.getController(component)
            val bounds = controller.getBounds(componentsFacility.getModelForm(component))

            val shift = Point()
            if (bounds.x > rightBound) {
                shift.translate(dx, 0)
            }
            if (bounds.y > bottomBound) {
                shift.translate(0, dy)
            }
            if (shift != Point()) {
                componentShifts[component] = shift
            }
        }
        return componentShifts
    }

    private fun createExpandedSceneCell(componentController: FunctionBlockController): EditorCell {
        val childInstance = componentController.fbInstance!!
        val node = (childInstance.declaration as PlatformElement).node
        val repository: SRepository = componentController.componentCell.context.repository
        val editorComponent = HeadlessEditorComponent(repository)
        editorComponent.updater.initialEditorHints = arrayOf(EXPANDED_FB_INSTANCE_KEY)
        editorComponent.editNode(node)
        val sceneCell: EditorCell = editorComponent.rootCell
        val style: Style = componentController.componentCell.style
        val fontSize = getFontSize(style)
        setFontSize(sceneCell.style, fontSize)
        sceneCell.relayout()
        return sceneCell
    }

    companion object {
        private const val EXPANDED_FB_INSTANCE_KEY =
            "org.fbme.ide.richediting.lang.editor.Rich Editing Hint.expanded_fb_instance"

        private const val EDITOR_SHIFT_X = 30
        private const val EDITOR_SHIFT_Y = 30
    }
}
